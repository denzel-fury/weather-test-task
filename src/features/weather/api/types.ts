export interface WeatherData {
  name: string;
  weather: {
    description: string;
    icon: string;
  }[];
  main: {
    temp: number;
    temp_min: number;
    temp_max: number;
  };
  wind: {
    speed: number;
  };
}

export interface CitySuggestion {
  name: string;
}

export interface SuggestionsApiResponse {
  list: CitySuggestion[];
}