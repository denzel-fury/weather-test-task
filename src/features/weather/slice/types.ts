
export interface WeatherState {
  history: string[];
  deletedItem: string | null;
}

export const initialState: WeatherState = {
  history: [],
  deletedItem: null,
};