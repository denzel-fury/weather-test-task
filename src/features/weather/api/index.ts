import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react';
import { WeatherData, SuggestionsApiResponse } from './types'
const API_KEY = "0c9231a730ccab5862b083452b48a013";

export const weatherApi = createApi({
  reducerPath: 'weatherApi',
  baseQuery: fetchBaseQuery({ baseUrl: 'https://api.openweathermap.org/data/2.5/' }),
  endpoints: (builder) => ({
    getWeather: builder.query<WeatherData, string>({
      query: (city) => `weather?q=${city}&appid=${API_KEY}&units=metric`,
    }),
    getSuggestions: builder.query<string[], string>({
      query: (query) => `find?q=${query}&type=like&appid=${API_KEY}&units=metric`,
      transformResponse: (res: SuggestionsApiResponse) => {
        const seen = new Set<string>();

        return res.list
          .map((item) => item.name)
          .filter((location) => {
            if (seen.has(location)) return false;
            seen.add(location);
            return true;
          });
      }
    }),
  }),
});

export const { useGetWeatherQuery, useGetSuggestionsQuery } = weatherApi;
