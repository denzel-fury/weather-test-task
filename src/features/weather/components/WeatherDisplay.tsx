import React from 'react'
import { useTranslation } from "react-i18next";


interface WeatherDisplayProps {
  weather: any;
}

const WeatherDisplay: React.FC<WeatherDisplayProps> = ({ weather }) => {
  const { t } = useTranslation();
  return (
    <div className="bg-gray-100 p-4 rounded mb-4" data-testid="weather-display">
      <h2 className="text-xl font-semibold">{weather.name}</h2>
      <p>{weather.weather[0].description}</p>
      <p>{t("temp")}: {weather.main.temp}°C</p>
      <p>{t("min")}: {weather.main.temp_min}°C / {t("max")}: {weather.main.temp_max}°C</p>
      <p>{t("wind")}: {weather.wind.speed} m/s</p>
    </div>
  );
};

const MemoizedWeatherDisplay = React.memo(WeatherDisplay);

export default MemoizedWeatherDisplay