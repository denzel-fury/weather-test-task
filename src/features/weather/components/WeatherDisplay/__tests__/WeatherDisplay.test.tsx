import React from 'react';
import { render, screen } from '@testing-library/react';
import WeatherDisplay from '../index';
import { WeatherData } from '../../../api/types'

jest.mock('react-i18next', () => ({
  useTranslation: () => ({
    t: (key: string) => key, // simple passthrough
  }),
}));

describe('WeatherDisplay', () => {
  const mockWeather: WeatherData = {
    name: 'Paris',
    weather: [{ description: 'clear sky', icon: 'mock' }],
    main: {
      temp: 25,
      temp_min: 20,
      temp_max: 28,
    },
    wind: {
      speed: 5,
    },
  };

  test('renders weather data correctly', () => {
    render(<WeatherDisplay weather={mockWeather} />);

    const display = screen.getByTestId('weather-display');
    expect(display).toBeInTheDocument();

    expect(screen.getByText('Paris')).toBeInTheDocument();
    expect(screen.getByText('clear sky')).toBeInTheDocument();
    expect(screen.getByText(/temp.*25°C/)).toBeInTheDocument();
    expect(screen.getByText(/min.*20°C.*max.*28°C/)).toBeInTheDocument();
    expect(screen.getByText(/wind.*5 m\/s/)).toBeInTheDocument();
  });

  test('matches snapshot', () => {
    const { asFragment } = render(<WeatherDisplay weather={mockWeather} />);;
    expect(asFragment()).toMatchSnapshot();
  });
});
