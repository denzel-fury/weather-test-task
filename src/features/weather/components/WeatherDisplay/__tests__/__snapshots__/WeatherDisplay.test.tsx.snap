// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`WeatherDisplay matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="bg-gray-100 p-4 rounded mb-4"
    data-testid="weather-display"
  >
    <h2
      class="text-xl font-semibold"
    >
      Paris
    </h2>
    <p>
      clear sky
    </p>
    <p>
      temp: 25°C
    </p>
    <p>
      min: 20°C / max: 28°C
    </p>
    <p>
      wind: 5 m/s
    </p>
  </div>
</DocumentFragment>
`;
